flink-streaming-java/src/test/java/org/apache/flink/streaming/util/AbstractStreamOperatorTestHarness.java

```diff
@@ -492,6 +503,10 @@ public void close() throws Exception {
 			processingTimeService.shutdownService();
 		}
 		setupCalled = false;
+
+		if (internalEnvironment.isPresent()) {
```

Comment:
Personal preference here, can ignore if you don't agree:

I somehow find it more intuitive to just say:
```
if (environmentIsInternal) {
    environment.close();
}
```
i.e., we keep the boolean field `environmentIsInternal` instead of an optional field called `internalEnvironment`, which I personally got a bit confused with the other `environment` (while they are actually the same environment instance).

Reply:
Or, another option, if we make sure the `MockEnvironment::close()` method is idempotent, we don't really need to differentiate whether or not the environment is internally created, and just always close the environment when closing the test harness.

That could probably make things even more safer?

Reply:
I think to enable this `Environment` must implement `AutoCloseable` as well. Maybe an empty default `close()` method? 
If you decide to stick with `Optional`, maybe change this line to: `internalEnvironment.ifPresent(MockEnvironment::close);`

Reply:
👍 Agreed.